File: backend/internal/service/service_darwin.go
//go:build darwin

package service

import (
	"context"
	"fmt"
	"log"
	"os"
	"os/exec"
	"os/signal"
	"path/filepath"
	"syscall"
	"text/template"
	"time"

	"github.com/lambertse/windows-app-locktime/backend/internal/api"
	"github.com/lambertse/windows-app-locktime/backend/internal/db"
	"github.com/lambertse/windows-app-locktime/backend/internal/watcher"
)

const (
	ServiceLabel   = "com.lambertse.applocker"
	DBPath         = "/Library/Application Support/AppLocker/applocker.db"
	ListenAddr     = "127.0.0.1:8089"
	FrontendAddr   = "127.0.0.1:8090"
	PlistPath      = "/Library/LaunchDaemons/com.lambertse.applocker.plist"
	LogDir         = "/var/log/applocker"
	installHtmlDir = "/usr/local/share/applocker/html"
)

// htmlDir returns the frontend dist directory. When installed via .pkg the
// files live at the fixed system path; during local development they are
// expected next to the binary.
func htmlDir() string {
	if _, err := os.Stat(installHtmlDir); err == nil {
		return installHtmlDir
	}
	exePath, _ := os.Executable()
	return filepath.Join(filepath.Dir(exePath), "html")
}

const launchdPlist = `<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>Label</key>
	<string>com.lambertse.applocker</string>
	<key>ProgramArguments</key>
	<array>
		<string>{{.ExePath}}</string>
		<string>--run</string>
	</array>
	<key>RunAtLoad</key>
	<true/>
	<key>KeepAlive</key>
	<true/>
	<key>StandardOutPath</key>
	<string>/var/log/applocker/applocker.log</string>
	<key>StandardErrorPath</key>
	<string>/var/log/applocker/applocker.error.log</string>
</dict>
</plist>
`

// RunService starts the API server, SPA server, and process watcher, then
// blocks until SIGTERM or SIGINT is received (managed by launchd or the user).
func RunService() error {
	startedAt := time.Now()

	if err := os.MkdirAll(filepath.Dir(DBPath), 0750); err != nil {
		return fmt.Errorf("mkdir db dir: %w", err)
	}

	database, err := db.Open(DBPath)
	if err != nil {
		return fmt.Errorf("open db: %w", err)
	}
	defer database.Close()

	if err := db.CrashRecovery(database, startedAt); err != nil {
		log.Printf("service: crash recovery: %v", err)
	}

	// Start API server (8089).
	apiServer := &api.Server{DB: database, StartedAt: startedAt}
	apiRouter := api.SetupRouter(apiServer)
	go func() {
		if err := apiRouter.Run(ListenAddr); err != nil {
			log.Printf("service: api server: %v", err)
		}
	}()

	// Start Go SPA server (8090).
	frontendRouter := api.NewFrontendRouter(htmlDir())
	go func() {
		if err := frontendRouter.Run(FrontendAddr); err != nil {
			log.Printf("service: frontend server: %v", err)
		}
	}()

	// Start process watcher.
	w := watcher.New(database)
	w.Start()

	log.Printf("service: applocker started (api=%s frontend=%s)", ListenAddr, FrontendAddr)

	// Block until the OS or launchd sends a stop signal.
	ctx, stop := signal.NotifyContext(context.Background(), syscall.SIGTERM, syscall.SIGINT)
	defer stop()
	<-ctx.Done()

	log.Printf("service: shutting down")
	w.CloseAllSessions("natural")
	return nil
}

// Install writes the LaunchDaemon plist and loads it via launchctl.
// Must be run as root.
func Install(exePath string) error {
	if err := os.MkdirAll(LogDir, 0755); err != nil {
		return fmt.Errorf("mkdir log dir: %w", err)
	}
	if err := os.MkdirAll(filepath.Dir(PlistPath), 0755); err != nil {
		return fmt.Errorf("mkdir plist dir: %w", err)
	}

	f, err := os.Create(PlistPath)
	if err != nil {
		return fmt.Errorf("create plist: %w", err)
	}
	defer f.Close()

	tmpl, err := template.New("plist").Parse(launchdPlist)
	if err != nil {
		return fmt.Errorf("parse plist template: %w", err)
	}
	if err := tmpl.Execute(f, map[string]string{"ExePath": exePath}); err != nil {
		return fmt.Errorf("write plist: %w", err)
	}

	if out, err := exec.Command("launchctl", "bootstrap", "system", PlistPath).CombinedOutput(); err != nil {
		return fmt.Errorf("launchctl bootstrap: %v: %s", err, out)
	}

	return nil
}

// Uninstall unloads and removes the LaunchDaemon. Must be run as root.
func Uninstall() error {
	// Best-effort unload; ignore error if already unloaded.
	_ = exec.Command("launchctl", "bootout", "system", PlistPath).Run()

	if err := os.Remove(PlistPath); err != nil && !os.IsNotExist(err) {
		return fmt.Errorf("remove plist: %w", err)
	}
	return nil
}

File: backend/src/service/service_darwin.cpp
#ifndef _WIN32
#include <ibridger/rpc/server.h>

#include <atomic>
#include <chrono>
#include <csignal>
#include <cstdio>
#include <cstring>
#include <fstream>
#include <memory>
#include <string>
#include <system_error>

#include "common/constants.h"
#include "common/utils.h"
#include "db/database.h"
#include "rpc/locktime_service.h"
#include "service_manager.h"
#include "watcher/watcher.h"

// macOS launchd plist path for system daemon
static constexpr const char* kLaunchdPlistPath =
    "/Library/LaunchDaemons/com.lambertse.locktime.plist";

namespace locktime {

// ── Signal handling
// ───────────────────────────────────────────────────────────

static std::atomic<bool> g_should_stop{false};

static void signal_handler(int /*sig*/) { g_should_stop = true; }

// ── run_service
// ───────────────────────────────────────────────────────────────

int run_service() {
  // Install signal handlers
  struct sigaction sa{};
  sa.sa_handler = signal_handler;
  sigemptyset(&sa.sa_mask);
  sa.sa_flags = 0;
  sigaction(SIGTERM, &sa, nullptr);
  sigaction(SIGINT, &sa, nullptr);

  std::shared_ptr<Database> db;
  std::unique_ptr<Watcher> watcher;
  std::unique_ptr<ibridger::rpc::Server> rpc_server;

  try {
    db = std::make_shared<Database>(kDbPath);
    db->crash_recovery(std::time(nullptr));

    auto started_at = std::chrono::steady_clock::now();
    auto svc = std::make_shared<LockTimeService>(db, started_at);

    ibridger::rpc::ServerConfig rpc_cfg;
    rpc_cfg.endpoint = kRpcEndpoint;
    rpc_server = std::make_unique<ibridger::rpc::Server>(rpc_cfg);
    rpc_server->register_service(svc);
    rpc_server->start();

    watcher = std::make_unique<Watcher>(db);
    watcher->start();
  } catch (const std::exception& ex) {
    std::fprintf(stderr, "locktime-svc: startup error: %s\n", ex.what());
    return 1;
  }

  // Event loop — sleep until signalled
  while (!g_should_stop) {
    struct timespec ts{1, 0};
    nanosleep(&ts, nullptr);
  }

  // Shutdown
  if (watcher) {
    watcher->close_all_sessions("service_stop");
    watcher->stop();
  }
  if (rpc_server) {
    rpc_server->stop();
  }

  return 0;
}

// ── install_service
// ───────────────────────────────────────────────────────────

std::error_code install_service(const std::string& exe_path) {
  // Write a launchd plist
  std::string plist =
      "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
      "<!DOCTYPE plist PUBLIC \"-//Apple//DTD PLIST 1.0//EN\"\n"
      "  \"http://www.apple.com/DTDs/PropertyList-1.0.dtd\">\n"
      "<plist version=\"1.0\">\n"
      "<dict>\n"
      "  <key>Label</key>\n"
      "  <string>" +
      std::string(kServiceName) +
      "</string>\n"
      "  <key>ProgramArguments</key>\n"
      "  <array>\n"
      "    <string>" +
      exe_path +
      "</string>\n"
      "    <string>--run</string>\n"
      "  </array>\n"
      "  <key>RunAtLoad</key><true/>\n"
      "  <key>KeepAlive</key><true/>\n"
      "  <key>StandardErrorPath</key>\n"
      "  <string>/Library/Logs/locktime.log</string>\n"
      "  <key>StandardOutPath</key>\n"
      "  <string>/Library/Logs/locktime.log</string>\n"
      "</dict>\n"
      "</plist>\n";

  std::ofstream f(kLaunchdPlistPath);
  if (!f) {
    return std::error_code(errno, std::generic_category());
  }
  f << plist;
  f.close();

  // Load via launchctl
  int rc =
      std::system(("launchctl load " + std::string(kLaunchdPlistPath)).c_str());
  if (rc != 0) {
    return std::error_code(rc, std::generic_category());
  }
  return {};
}

// ── uninstall_service
// ─────────────────────────────────────────────────────────

std::error_code uninstall_service() {
  // Unload from launchd
  std::system(("launchctl unload " + std::string(kLaunchdPlistPath)).c_str());

  // Remove plist file
  if (std::remove(kLaunchdPlistPath) != 0 && errno != ENOENT) {
    return std::error_code(errno, std::generic_category());
  }
  return {};
}

}  // namespace locktime
#endif  // !_WIN32
